indexers/elastic_test.go

```diff
@@ -0,0 +1,132 @@
+package indexers
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"os"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Tests for elastic.go", func() {
+	Context("Tests for new()", func() {
+		payload := []byte(`{
+			"name" : "0bcd132328f2f0c8ee451d471960750e",
+			"cluster_name" : "415909267177:perfscale-dev",
+			"cluster_uuid" : "Xz2IU4etSieAeaO2j-QCUw",
+			"version" : {
+			  "number" : "7.10.2",
+			  "build_type" : "tar",
+			  "build_hash" : "unknown",
+			  "build_date" : "2023-03-22T14:16:51.874273Z",
+			  "build_snapshot" : false,
+			  "lucene_version" : "9.3.0",
+			  "minimum_wire_compatibility_version" : "7.10.0",
+			  "minimum_index_compatibility_version" : "7.0.0"
+			},
+			"tagline" : "The OpenSearch Project: https://opensearch.org/"
+		  }`)
+		testcase := struct {
+			indexerConfig IndexerConfig
+			mockServer    *httptest.Server
+		}{
+			indexerConfig: IndexerConfig{Type: "elastic",
+				Servers:            []string{},
+				Index:              "go-commons-test",
+				InsecureSkipVerify: true,
+			},
+			mockServer: httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusOK)
+				w.Write(payload)
+			})),
+		}
+
+		It("Returns error status bad request", func() {
+			var indexer Elastic
+			indexer.index = "go-commons-test"
+			testcase.mockServer = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusBadRequest)
+			}))
+			defer testcase.mockServer.Close()
+			testcase.indexerConfig.Servers = []string{testcase.mockServer.URL}
+			err := indexer.new(testcase.indexerConfig)
+			Expect(err).NotTo(BeNil())
+		})
+
+		It("when no url is passed", func() {
+			var indexer Elastic
+			indexer.index = "go-commons-test"
+			err := indexer.new(testcase.indexerConfig)
+			testcase.mockServer = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusGatewayTimeout)
+			}))
+			Expect(err).NotTo(BeNil())
+		})
+
+		It("Returns err not passing a valid URL in env variable", func() {
+			var indexer Elastic
+			indexer.index = "go-commons-test"
+			testcase.indexerConfig.Servers = []string{}
+			os.Setenv("ELASTICSEARCH_URL", "not a valid url:port")
+			defer os.Unsetenv("ELASTICSEARCH_URL")
+			defer testcase.mockServer.Close()
+			err := indexer.new(testcase.indexerConfig)
+			Expect(err).NotTo(BeNil())
+		})
+
+		It("Returns err no index name", func() {
+			var indexer Elastic
+			indexer.index = "go-commons-test"
+			defer testcase.mockServer.Close()
+			testcase.indexerConfig.Servers = []string{testcase.mockServer.URL}
+			testcase.indexerConfig.Index = ""
+			err := indexer.new(testcase.indexerConfig)
+			Expect(err).NotTo(BeNil())
+		})
+
+	})
+
+	Context("Tests for Index()", func() {
```

Comment:
For indexing I think we can have a success case by re-index this doc every time which should be ok as it keeps on updating the same doc everytime.[Same applies for opensearch.go as well]
```
{
  "_index": "dummy",
  "_id": "Vdc8g4IBYDIaJD7xG1yv",
  "_version": 1,
  "_score": null,
  "_source": {
    "timestamp": "2022-08-09T15:32:10.784425",
    "uuid": "de4f28de-93de-4ab1-90dc-bdd67653b895",
    "connection_time": 0.00347347604110837
  },
  "fields": {
    "timestamp": [
      "2022-08-09T15:32:10.784Z"
    ]
  },
  "sort": [
    1660059130784
  ]
}
```